### ISSUE TITLE
Graal gives different result from the interpreter

### ISSUE THREAD

**Issue by connglli** [oracle/graal#4754]
**Describe the issue**

This is a JIT bug where Graal gives a different output from the interpreter or HotSpot's C1/C2 compiler.

**Steps to reproduce the issue**

1. javac T.java
2. java -Xint T >aaa
3. java T >bbb
4. diff aaa bbb

The results of aaa and bbb are different, while expected to be the same.

**Describe GraalVM and your environment:**
 - GraalVM version: 22.3.0-dev
 - JDK major version: 11.0.16
 - OS: Ubuntu 20.04
 - Architecture: x86_64

**More details**

The following is the source of our tests:

``` java
class T {
  long r;

  void a() {
    int x;
    long y;
    for (x = 3; x > 1; x--)
      for (y = 1; 2 > y; y++) {
        r += y;
        r &= x;
      }
  }

  void b(long l) {
    for (int z = 0; z < 12712; z += 1) a();
    a();
  }

  void f() {
    b(r);
    System.out.println("r = " + r);
  }

  public static void main(String[] g) {
    T t = new T();
    for (int i = 0; i < 10; i++) t.f();
  }
}
```

Result of the interpreter:

```
$ java -Xint T
r = 2
r = 0
r = 2
r = 0
r = 2
r = 0
r = 2
r = 0
r = 2
r = 0
```

Result of Graal compiler:

```
$ java T
r = 2
r = 2
r = 2
r = 2
r = 2
r = 2
r = 2
r = 2
r = 2
r = 2
```

**Comment by davleopo** [comment]
@connglli  Thanks for the report, I'll have a look.

**Comment by gergo-** [comment]
@connglli this should be fixed by @andrewcraik 's 169807f. Please check if the newest development build from https://github.com/graalvm/graalvm-ce-dev-builds/releases behaves as expected.

**Comment by connglli** [comment]
Close this issue as fixeded by https://github.com/oracle/graal/commit/169807f482108026555bb893f96ee7dd18d64ea6
